File: DreamPortalSelenium/src/test/java/com/dreamportal/pages/DreamTotalPage.java
package com.dreamportal.pages;

import com.dreamportal.utils.BasePage;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Dream Total (Summary) Page Object Model
 * 
 * This page object represents the Dream Summary page (dreams-total.html).
 * It handles all interactions and validations for dream statistics.
 * 
 * Responsibilities:
 * - Verify correct statistics (Good: 6, Bad: 4, Total: 10, Recurring: 2)
 * - Validate recurring dreams
 * - Extract and validate statistics displayed on the page
 */
public class DreamTotalPage extends BasePage {
    private static final Logger logger = LoggerFactory.getLogger(DreamTotalPage.class);

    // Web Element Locators
    private final By goodDreamsCount = By.xpath("//*[contains(text(), 'Good')] | //*[contains(text(), 'good')]");
    private final By badDreamsCount = By.xpath("//*[contains(text(), 'Bad')] | //*[contains(text(), 'bad')]");
    private final By totalDreamsCount = By.xpath("//*[contains(text(), 'Total')] | //*[contains(text(), 'total')]");
    private final By recurringDreamsCount = By.xpath("//*[contains(text(), 'Recurring')] | //*[contains(text(), 'recurring')]");
    private final By recurringDreamsList = By.xpath("//li[contains(text(), 'Flying')] | //li[contains(text(), 'Lost')]");

    /**
     * Constructor
     * 
     * @param driver WebDriver instance
     */
    public DreamTotalPage(WebDriver driver) {
        super(driver);
        logger.info("DreamTotalPage initialized");
    }

    /**
     * Navigate to dream total page
     * 
     * @param url URL of the dream total page
     */
    public void navigateToDreamTotal(String url) {
        logger.info("Navigating to dream total page: " + url);
        navigateTo(url);
        logger.info("Dream total page loaded");
    }

    /**
     * Extract number from text using regex
     * 
     * @param text Text containing the number
     * @return Extracted number
     */
    private int extractNumberFromText(String text) {
        Pattern pattern = Pattern.compile("\\d+");
        Matcher matcher = pattern.matcher(text);
        
        if (matcher.find()) {
            return Integer.parseInt(matcher.group());
        }
        return 0;
    }

    /**
     * Get good dreams count
     * 
     * @return Number of good dreams
     */
    public int getGoodDreamsCount() {
        logger.info("Getting good dreams count");
        try {
            List<WebElement> elements = driver.findElements(goodDreamsCount);
            for (WebElement element : elements) {
                String text = element.getText();
                if (text.toLowerCase().contains("good")) {
                    int count = extractNumberFromText(text);
                    logger.info("Good dreams count: " + count);
                    return count;
                }
            }
        } catch (Exception e) {
            logger.error("Error getting good dreams count: " + e.getMessage());
        }
        return 0;
    }

    /**
     * Get bad dreams count
     * 
     * @return Number of bad dreams
     */
    public int getBadDreamsCount() {
        logger.info("Getting bad dreams count");
        try {
            List<WebElement> elements = driver.findElements(badDreamsCount);
            for (WebElement element : elements) {
                String text = element.getText();
                if (text.toLowerCase().contains("bad")) {
                    int count = extractNumberFromText(text);
                    logger.info("Bad dreams count: " + count);
                    return count;
                }
            }
        } catch (Exception e) {
            logger.error("Error getting bad dreams count: " + e.getMessage());
        }
        return 0;
    }

    /**
     * Get total dreams count
     * 
     * @return Total number of dreams
     */
    public int getTotalDreamsCount() {
        logger.info("Getting total dreams count");
        try {
            List<WebElement> elements = driver.findElements(totalDreamsCount);
            for (WebElement element : elements) {
                String text = element.getText();
                if (text.toLowerCase().contains("total")) {
                    int count = extractNumberFromText(text);
                    logger.info("Total dreams count: " + count);
                    return count;
                }
            }
        } catch (Exception e) {
            logger.error("Error getting total dreams count: " + e.getMessage());
        }
        return 0;
    }

    /**
     * Get recurring dreams count
     * 
     * @return Number of recurring dreams
     */
    public int getRecurringDreamsCount() {
        logger.info("Getting recurring dreams count");
        try {
            List<WebElement> elements = driver.findElements(recurringDreamsCount);
            for (WebElement element : elements) {
                String text = element.getText();
                if (text.toLowerCase().contains("recurring")) {
                    int count = extractNumberFromText(text);
                    logger.info("Recurring dreams count: " + count);
                    return count;
                }
            }
        } catch (Exception e) {
            logger.error("Error getting recurring dreams count: " + e.getMessage());
        }
        return 0;
    }

    /**
     * Get all statistics at once
     * 
     * @return Map containing all statistics
     */
    public Map<String, Integer> getAllStats() {
        logger.info("Getting all statistics");
        
        Map<String, Integer> stats = new HashMap<>();
        stats.put("good", getGoodDreamsCount());
        stats.put("bad", getBadDreamsCount());
        stats.put("total", getTotalDreamsCount());
        stats.put("recurring", getRecurringDreamsCount());
        
        logger.info("Statistics - Good: " + stats.get("good") + 
                   ", Bad: " + stats.get("bad") + 
                   ", Total: " + stats.get("total") + 
                   ", Recurring: " + stats.get("recurring"));
        
        return stats;
    }

    /**
     * Check if a specific dream is recurring
     * 
     * @param dreamName Name of the dream
     * @return boolean true if dream is recurring
     */
    public boolean isRecurringDream(String dreamName) {
        logger.info("Checking if '" + dreamName + "' is recurring");
        
        List<String> recurringNames = getRecurringDreamNames();
        boolean isRecurring = recurringNames.stream()
            .anyMatch(name -> name.toLowerCase().contains(dreamName.toLowerCase()));
        
        logger.info("Is '" + dreamName + "' recurring: " + isRecurring);
        return isRecurring;
    }

    /**
     * Get all recurring dream names
     * 
     * @return List of recurring dream names
     */
    public List<String> getRecurringDreamNames() {
        logger.info("Getting all recurring dream names");
        
        List<String> names = new ArrayList<>();
        try {
            List<WebElement> elements = driver.findElements(recurringDreamsList);
            for (WebElement element : elements) {
                String name = element.getText().trim();
                names.add(name);
                logger.info("Found recurring dream: " + name);
            }
        } catch (Exception e) {
            logger.warn("Could not extract recurring dream names: " + e.getMessage());
        }
        
        return names;
    }

    /**
     * Validate expected recurring dreams are present
     * 
     * @return boolean true if both expected dreams are recurring
     */
    public boolean validateExpectedRecurringDreams() {
        logger.info("Validating expected recurring dreams");
        
        boolean hasFlying = isRecurringDream("Flying over mountains");
        boolean hasLost = isRecurringDream("Lost in maze");
        
        boolean isValid = hasFlying && hasLost;
        logger.info("Expected recurring dreams valid: " + isValid);
        
        return isValid;
    }

    /**
     * Get page title
     * 
     * @return Page title
     */
    public String getPageTitle() {
        logger.info("Getting dream total page title");
        return driver.getTitle();
    }

    /**
     * Verify page loaded successfully
     * 
     * @return boolean true if page is loaded
     */
    public boolean isPageLoaded() {
        logger.info("Checking if dream total page is loaded");
        return getTotalDreamsCount() > 0;
    }
}

File: DreamPortal/tests/pages/dreamTotalPage.ts
/**
 * Dream Total Page Object Model
 * 
 * This file contains the page object for the Dream Summary page (dreams-total.html).
 * It handles all interactions and validations for dream statistics.
 * 
 * Responsibilities:
 * - Verify the correct statistics: Good Dreams (6), Bad Dreams (4), Total Dreams (10)
 * - Validate Recurring Dreams count (2)
 * - Verify that "Flying over mountains" and "Lost in maze" are counted as recurring
 * - Extract and validate all statistics displayed on the page
 */

import { Page, Locator } from '@playwright/test';

export interface DreamStats {
  goodDreams: number;
  badDreams: number;
  totalDreams: number;
  recurringDreams: number;
  recurringDreamNames?: string[];
}

export class DreamTotalPage {
  readonly page: Page;
  readonly goodDreamsCount: Locator;
  readonly badDreamsCount: Locator;
  readonly totalDreamsCount: Locator;
  readonly recurringDreamsCount: Locator;
  readonly recurringDreamsList: Locator;

  constructor(page: Page) {
    this.page = page;
    
    // Define selectors for dream statistics elements
    // The page has a table with statistics where each row has Dream Type and Total Count columns
    // We locate the row containing the text, then get the cell value from the second column
    this.goodDreamsCount = page.locator('table tbody tr').filter({ hasText: /Good Dreams/i }).locator('td:nth-child(2)');
    this.badDreamsCount = page.locator('table tbody tr').filter({ hasText: /Bad Dreams/i }).locator('td:nth-child(2)');
    this.totalDreamsCount = page.locator('table tbody tr').filter({ hasText: /Total Dreams/i }).locator('td:nth-child(2)');
    this.recurringDreamsCount = page.locator('table tbody tr').filter({ hasText: /Recurring Dreams/i }).locator('td:nth-child(2)');
    this.recurringDreamsList = page.locator('[class*="recurring"] li, [id*="recurring"] li, [data-testid*="recurring"] li');
  }

  /**
   * Navigate to the dream total page
   * @param url - The URL of the dream total/summary page
   */
  async navigateToDreamTotal(url: string): Promise<void> {
    await this.page.goto(url);
  }

  /**
   * Get the number of good dreams displayed
   * @returns number - Count of good dreams
   */
  async getGoodDreamsCount(): Promise<number> {
    return await this.extractNumberFromText(this.goodDreamsCount);
  }

  /**
   * Get the number of bad dreams displayed
   * @returns number - Count of bad dreams
   */
  async getBadDreamsCount(): Promise<number> {
    return await this.extractNumberFromText(this.badDreamsCount);
  }

  /**
   * Get the total number of dreams displayed
   * @returns number - Total count of dreams
   */
  async getTotalDreamsCount(): Promise<number> {
    return await this.extractNumberFromText(this.totalDreamsCount);
  }

  /**
   * Get the number of recurring dreams displayed
   * @returns number - Count of recurring dreams
   */
  async getRecurringDreamsCount(): Promise<number> {
    return await this.extractNumberFromText(this.recurringDreamsCount);
  }

  /**
   * Helper method to extract numbers from text content
   * @param locator - The element locator containing the number
   * @returns number - The extracted number
   */
  private async extractNumberFromText(locator: Locator): Promise<number> {
    try {
      const text = await locator.textContent();
      if (!text) return 0;
      
      // Extract the first number found in the text
      const match = text.match(/\d+/);
      return match ? parseInt(match[0], 10) : 0;
    } catch (error) {
      console.warn('Error extracting number from text:', error);
      return 0;
    }
  }

  /**
   * Get all recurring dream names
   * @returns string[] - Array of recurring dream names
   */
  async getRecurringDreamNames(): Promise<string[]> {
    try {
      const count = await this.recurringDreamsList.count();
      const names: string[] = [];

      for (let i = 0; i < count; i++) {
        const name = await this.recurringDreamsList.nth(i).textContent();
        if (name) {
          names.push(name.trim());
        }
      }

      return names;
    } catch (error) {
      // If recurring dreams list is not found on the page, return empty array
      console.warn('Recurring dreams list not found on page:', error);
      return [];
    }
  }

  /**
   * Get all dream statistics at once
   * @returns DreamStats - Object containing all statistics
   */
  async getAllStats(): Promise<DreamStats> {
    return {
      goodDreams: await this.getGoodDreamsCount(),
      badDreams: await this.getBadDreamsCount(),
      totalDreams: await this.getTotalDreamsCount(),
      recurringDreams: await this.getRecurringDreamsCount(),
      recurringDreamNames: await this.getRecurringDreamNames()
    };
  }

  /**
   * Validate that a specific dream is in the recurring dreams list
   * Note: This method looks for the dream name on the page if it's displayed,
   * otherwise returns false (the dream might be recurring in data but not displayed on summary)
   * @param dreamName - The name of the dream to find
   * @returns boolean - True if the dream is in the recurring list displayed on page
   */
  async isRecurringDream(dreamName: string): Promise<boolean> {
    try {
      const recurringNames = await this.getRecurringDreamNames();
      if (recurringNames.length === 0) {
        // Recurring dreams list not displayed on page
        // Return false - the validation should be done through data comparison
        return false;
      }
      return recurringNames.some(name => 
        name.toLowerCase().includes(dreamName.toLowerCase())
      );
    } catch (error) {
      console.warn(`Error checking if "${dreamName}" is recurring:`, error);
      return false;
    }
  }

  /**
   * Validate that the expected recurring dreams are present
   * @returns boolean - True if both "Flying over mountains" and "Lost in maze" are recurring
   */
  async validateExpectedRecurringDreams(): Promise<boolean> {
    const isFlying = await this.isRecurringDream('Flying over mountains');
    const isLost = await this.isRecurringDream('Lost in maze');
    return isFlying && isLost;
  }

  /**
   * Calculate recurring dreams from dream entries
   * A dream is recurring if it appears more than once
   * @param dreams - Array of dream entries
   * @returns string[] - Array of unique recurring dream names
   */
  getRecurringDreamsFromData(dreams: any[]): string[] {
    const dreamCounts: { [key: string]: number } = {};
    
    // Count occurrences of each dream name
    dreams.forEach(dream => {
      dreamCounts[dream.name] = (dreamCounts[dream.name] || 0) + 1;
    });
    
    // Return dreams that appear more than once
    return Object.keys(dreamCounts).filter(name => dreamCounts[name] > 1);
  }

  /**
   * Verify that expected recurring dreams exist in the given dream data
   * @param dreams - Array of dream entries
   * @returns boolean - True if both expected recurring dreams are found
   */
  validateExpectedRecurringDreamsInData(dreams: any[]): boolean {
    const recurringDreams = this.getRecurringDreamsFromData(dreams);
    
    const hasFlying = recurringDreams.some(name =>
      name.toLowerCase().includes('flying') && name.toLowerCase().includes('mountain')
    );
    const hasLost = recurringDreams.some(name =>
      name.toLowerCase().includes('lost') && name.toLowerCase().includes('maze')
    );
    
    return hasFlying && hasLost;
  }
}
